### PR TITLE
Add TFileCacheWrite and make cacheSize an option for ROOTOutputer

### PR DESCRIPTION
**Pull request by gartung**
Resubmit of https://github.com/hep-cce2/root_serialization/pull/42 for code review after resetting head of master

### CODE REVIEW

**Comment by Dr15Jones** [comment]
Could you change the description of the pull request to say what this is?

File: RootOutputer.cc [hunk 1]
```diff
@@ -20,6 +21,7 @@ RootOutputer::RootOutputer(std::string const& iFileName, unsigned int iNLanes, C
   basketSize_{iConfig.basketSize_},
   splitLevel_{iConfig.splitLevel_}
 {
+  TFileCacheWrite(&file_, iConfig.cacheSize_);
```
Comment (Dr15Jones) [inline RootOutputer.cc:24]: This has the `new` missing.

File: RootOutputerConfig.h [hunk 2]
```diff
@@ -13,6 +13,7 @@ namespace cce::tf {
     int basketSize_=16384;
     int treeMaxVirtualSize_=-1;
     int autoFlush_=-1;
+    int cacheSize_=102400;
```
Comment (Dr15Jones) [inline RootOutputerConfig.h:16]: I believe this type is used by all the ROOT*Outputer as well. Please update the others to use this.

File: RootOutputer.cc (hunk 1 above)
Comment (Dr15Jones) [reply RootOutputer.cc:24]: Also, please make this optional by using a new configuration option to turn this one. The default of the new option should be so this is off.

Comment (Dr15Jones) [reply RootOutputer.cc:24]: Actually, just using `cacheSize_>0` is good enough. No need for another configuration parameter.

Comment (gartung) [reply RootOutputer.cc]: Sorry. I pushed from the mirror again.

File: RootOutputerConfig.h (hunk 2 above)
Comment (gartung) [reply RootOutputerConfig.h]: I didn't see any use of RootOutputConfig in the others. Do they get it through inheritance?

**Comment by gartung** [comment]
```
gartung@cori09:~/cscratch/cce-ios/root_serialization> git grep RootOutputerConfig.h
RootOutputer.cc:#include "RootOutputerConfig.h"
RootOutputerConfig.cc:#include "RootOutputerConfig.h"
RootOutputerConfig.h:#if !defined(RootOutputerConfig_h)
RootOutputerConfig.h:#define RootOutputerConfig_h
TBufferMergerRootOutputer.cc:#include "RootOutputerConfig.h"
```

**Comment by gartung** [comment]
I force pushed from the correct directory so your comments were for outdated code.

**Comment by Dr15Jones** [comment]
It also seems to be used here
https://github.com/Dr15Jones/root_serialization/blob/e8c542579333b29b55d5447cbf3832de89c156fd/TBufferMergerRootOutputer.cc#L213

File: TBufferMergerRootOutputer.h [hunk 3]
```diff
@@ -26,6 +26,7 @@ class TBufferMergerRootOutputer :public OutputerBase {
     int compressionLevel_=9;
     std::string compressionAlgorithm_="";
     int basketSize_=16384;
+    int cacheSize_=0;
```
Comment (gartung) [inline TBufferMergerRootOutputer.h:29]: @Dr15Jones I added it here.

**Comment by Dr15Jones** [comment]
In `TBufferMergerRootOutputer::TBufferMergerRootOutputer(...)` I think we might also need
```C++
     if(iConfig.cacheSize_ > 0 ) { 
      new TFileCacheWrite(buffer_.GetFile(), iConfig.cacheSize_);
   }
```
@pcanal what do you think?

**Comment by pcanal** [comment]
I think you are right.

**Comment by gartung** [comment]
Not quite
```
root-6.26.00-5bqqtsrl2kjwjawszrj5dqyking2ykyl/include/TFileCacheWrite.h:35:27: note:   no known conversion for argument 1 from 'std::shared_ptr<ROOT::TBufferMergerFile>' to 'TFile*'
   35 |    TFileCacheWrite(TFile *file, Int_t buffersize);
      |                    ~~~~~~~^~~~
```

File: TBufferMergerRootOutputer.cc [hunk 4]
```diff
@@ -41,6 +42,9 @@ TBufferMergerRootOutputer::TBufferMergerRootOutputer(std::string const& iFileNam
                     autoFlush_{iConfig.autoFlush_ != -1 ? iConfig.autoFlush_ : Config::kDefaultAutoFlush },
                     concurrentWrite_{iConfig.concurrentWrite}
 {
+   if(iConfig.cacheSize_ > 0 ) {
+      new TFileCacheWrite(buffer_.GetFile(), iConfig.cacheSize_);
```
Comment (Dr15Jones) [inline TBufferMergerRootOutputer.cc:46]:
```suggestion
      new TFileCacheWrite(buffer_.GetFile().get(), iConfig.cacheSize_);
```
?

Comment (gartung) [reply TBufferMergerRootOutputer.cc:46]: This worked
```
new TFileCacheWrite(buffer_.GetFile()->CurrentFile(), iConfig.cacheSize_);
```

File: TBufferMergerRootOutputer.cc [hunk 5]
```diff
@@ -41,6 +42,9 @@ TBufferMergerRootOutputer::TBufferMergerRootOutputer(std::string const& iFileNam
                     autoFlush_{iConfig.autoFlush_ != -1 ? iConfig.autoFlush_ : Config::kDefaultAutoFlush },
                     concurrentWrite_{iConfig.concurrentWrite}
 {
+   if(iConfig.cacheSize_ > 0 ) {
+      new TFileCacheWrite(buffer_.GetFile()->CurrentFile(), iConfig.cacheSize_);
+   }
```
Comment (Dr15Jones) [inline TBufferMergerRootOutputer.cc:47]: @pcanal is this what is needed?

Comment (pcanal) [reply TBufferMergerRootOutputer.cc:47]: Actually no, if I read correctly, that file is the `TMemFile` so the Write cache is useless.

**Comment by pcanal** [comment]
It appears the output file is not currently directly accessible.  2 options.
In https://github.com/hep-cce2/root_serialization/blob/1d7786a93b4d47155f611d1ef57f4cb02fee1a6e/TBufferMergerRootOutputer.cc#L37
you could create the TFile explicitly and then call/use:
```TBufferMerger::TBufferMerger(std::unique_ptr<TFile> output)```
or you can retrieve the file from:
```
gROOT->GetListOfFiles(iFileName.c_str());
```

**Comment by Dr15Jones** [comment]
I'd prefer 
```
TBufferMerger::TBufferMerger(std::unique_ptr<TFile> output)
```

**Comment by gartung** [comment]
So should I add a new class member `std::unique_ptr<TFile> output` and padd that to the `buffer_` inializer?

**Comment by Dr15Jones** [comment]
Hmm, given the current data structure, how about just doing Philippe's other suggestion:
```
gROOT->GetListOfFiles(iFileName.c_str());
```

**Comment by pcanal** [comment]
You could consider
```
std::unique_ptr<TFile> CreateFile(const char *filename, const char *option) {
   auto file = TFile::Open(filename, option);
   new TFileCacheWrite(file, ....);
   return unique_ptr<TFile>(file);
}
```
(after correcting the syntax errors :) )

**Comment by Dr15Jones** [comment]
If we change `CreateFile` to `createFile` and make it a private static member of TBufferMergerRootOutputer then I'm all for Philippe's suggestion.

**Comment by gartung** [comment]
I get an error with `ROOT::CompressionSettings`
```
root_serialization/TBufferMergerRootOutputer.h:64:86: error: 'ROOT::CompressionSettings' is not a type
```
The documentation says it is deprecated.

https://root.cern/doc/master/namespaceROOT.html#aebd7cf84a018a1135047edc0258f8d06

Should I use `RCompressionSetting` instead?

https://root.cern/doc/master/structROOT_1_1RCompressionSetting.html

**Comment by pcanal** [comment]
You should pass `RCompressionSetting::EAlgorithm algorithm, int compressionLevel`  and continue to use the function `ROOT::CompressionSettings` to create the int passed on those two values.

**Comment by gartung** [comment]
Sorry I need to rename this repo something other than origin. I keep accidentally pushing master to it.

**Pull request merged**